Add support for custom TLVs

commodo opened the pull request:
Added commands for `lldpcli`:
- `configure lldp custom-tlv oui 88,99,AA, subtype BB <oui-info CC,DD,EE...>` - to add a single custom TLV
- `unconfigure lldp custom-tlvs` - to remove all custom TLVs set

Add support for lldpctl to display custom TLVs.
Something like below [some info was redacted out].
```
Interface:    lan2, via: LLDP, RID: 3, Time: 0 day, 00:00:03
  Chassis:     
    ChassisID:    mac 74:D4:35:0A:C7:F3
    SysName:      192.168.0.23
    SysDescr:     OpenWRT Chaos Chalmer
  Port:        
    PortID:       local BLUE_LAN
    PortDescr:    eth0
  CustomTLVs:  
    TLV:          OUI: 11,22,33, SubType: 44, Len: 0
    TLV:          OUI: 11,22,33, SubType: 44, Len: 0
    TLV:          OUI: 11,22,33, SubType: 44, Len: 4 22,33,44,55
    TLV:          OUI: 11,22,33, SubType: 44, Len: 3 77,88,99
```

No hurry to review this.
Will also wait for build errors that may be caught by Travis-CI,

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo commented:
travis seems fine now

vincentbernat commented:
So, almost all good for me. I have put some inline comments with corrections.

Thanks!

vincentbernat commented:
Also, the support should be optional (like fot Dot1/Dot3). Some people want to keep lldpd size down. But this can be fixed later.

commodo commented:
I'll update my PR based on your review in the next couple of days.
Thanks for taking the time to do the review.

I'll add the commits on top of these, and discuss later whether we squash them within the current commits.
There's no rush at the moment, since we're currently using a patched version of 0.7.14 (with some of my patches) in our tree.

Regarding making custom-TLVs optional, I agree.
A question: would it be an idea to leave some that built-in into `lldpd` ?
But only the part where users would see unknown TLVs in `lldpctl`, rather than being (somewhat) silently discarded.

vincentbernat commented:
I only worry about what is in `lldpd`. The unknown TLV in `lldpctl` can stay. In `lldpd`, they have to be optional but enabled by default.

commodo commented:
I have a bit of a problem with the last commit, where I convert `oui_info` to a dynamic array with `MARSHAL_FSTR`.

The stack-trace is:
```
#0  __GI_memcpy (to=to@entry=0x480d4010, from=from@entry=0x106bf138, len=<optimized out>, len@entry=1048576) at libc/string/powerpc/memcpy.c:43
#1  0x4801e104 in marshal_serialize_ (mi=0x48031340 <marshal_info_fstring>, unserialized=unserialized@entry=0x106bf138, input=input@entry=0xbfe7cb7c, skip=0, _refs=_refs@entry=0x106bf270, 
    osize=1048576) at marshal.c:134
#2  0x4801e1d8 in marshal_serialize_ (mi=0x480316a8 <marshal_info_lldpd_custom>, unserialized=unserialized@entry=0x106bed50, input=input@entry=0xbfe7cbec, skip=<optimized out>, 
    _refs=_refs@entry=0x106bf270, osize=1048576) at marshal.c:153
#3  0x4801e1d8 in marshal_serialize_ (mi=mi@entry=0x480315b8 <marshal_info_lldpd_port_set>, unserialized=unserialized@entry=0xbfe7ccd0, input=input@entry=0xbfe7cc60, skip=skip@entry=0, 
    _refs=_refs@entry=0x0, osize=osize@entry=0) at marshal.c:153
#4  0x4801f350 in ctl_msg_send_unserialized (output_buffer=output_buffer@entry=0x106bda0c, output_len=output_len@entry=0x106bda14, type=type@entry=SET_PORT, t=t@entry=0xbfe7ccd0, 
    mi=0x480315b8 <marshal_info_lldpd_port_set>) at ctl.c:137
#5  0x4801b350 in _lldpctl_do_something (conn=0x106bd9f8, state_send=state_send@entry=5, state_recv=state_recv@entry=6, state_data=0x106bf508 "50020x106bf4b0lan0", 
    type=type@entry=SET_PORT, to_send=to_send@entry=0xbfe7ccd0, mi_send=<optimized out>, to_recv=to_recv@entry=0x0, mi_recv=mi_recv@entry=0x0) at atom.c:319
#6  0x4801ccc0 in _lldpctl_atom_set_atom_port (atom=0x106bf408, key=<optimized out>, value=<optimized out>) at atoms/port.c:383
#7  0x10007ea0 in cmd_custom_tlv_set (conn=0x106bd9f8, w=<optimized out>, env=0x0, arg=<optimized out>) at conf-lldp.c:278
#8  0x1000a834 in _commands_execute (conn=conn@entry=0x106bd9f8, w=w@entry=0x106bea50, root=root@entry=0x106bd068, argc=argc@entry=9, argv=argv@entry=0x106bc060, word=word@entry=0x0, 
    all=all@entry=0, priv=1) at commands.c:424
#9  0x1000a980 in commands_execute (conn=conn@entry=0x106bd9f8, w=w@entry=0x106bea50, root=root@entry=0x106bd068, argc=argc@entry=9, argv=argv@entry=0x106bc060, privileged=<optimized out>)
    at commands.c:553
#10 0x10002264 in cmd_exec (conn=conn@entry=0x106bd9f8, fmt=fmt@entry=0x1000d400 "plain", argc=9, argv=0x106bc060) at lldpcli.c:283
#11 0x1000237c in parse_and_exec (conn=conn@entry=0x106bd9f8, fmt=fmt@entry=0x1000d400 "plain", 
    line=line@entry=0x106bf0c8 "configure lldp custom-tlv oui 11,22,33 subtype 00 oui-info 44,55,66,77,88,99,AA") at lldpcli.c:318
#12 0x10001854 in main (argc=9, argv=0xbfe7d3d4) at lldpcli.c:49
```

Any idea ?
Maybe I have to add a few more marshals somewhere ?

vincentbernat commented:
So, try by replacing `uint16_t` by `int` and if this still doesn't work, I will dig a bit more. But otherwise, this seems correct.

commodo commented:
So, I updated both your recommendations and it worked.
Thanks :)

Any other notes/objections ? 
If it's ok with you, I think I'll squash some commits within the previous set of commits (where it makes sense).
I'll leave it up to you to decide if I do this.

vincentbernat commented:
❦ 18 mai 2015 08:43 -0700, Alexandru Ardelean <notifications@github.com> :

> So, I updated both your recommendations and it worked.
> Thanks :)
>
> Any other notes/objections ? 
> If it's ok with you, I think I'll squash some commits within the
> previous set of commits (where it makes sense).
> I'll leave it up to you to decide if I do this.

Yes, you can squash. Keep the commit which changes oui_info to a dynamic
array into a separate commit. It will be easier to bissect if we
discover a regression.
-- 
Make the coupling between modules visible.
            - The Elements of Programming Style (Kernighan & Plauger)

commodo commented:
So, I've been squashing commits when I was a bit more silent.
And testing the code.

I went ahead with `_lldpctl_alloc_in_atom()`, rebased on the latest master, and squashed some commits.
Please take a look and see if it's fine.

I do have 1 single request, if it's not too much trouble (and when you have time).
It is optional.
I've been trying to add a `--disable-stack-protector` option, but it took me too long to do it, so I gave up
On OpenWRT, it's disabled by default (not sure why, because I do not maintain `lldpd` on OpenWRT, that's someone else). It could be that the toolchain (uClibc or musl) do not support stack-protection.
Anyway, could you add it before releasing 0.7.15 ?
I've tried `lldp_ARG_ENABLE` but it did not seem to work for me.

Right now, on OpenWRT, there's a patch that just removes the 2 stack-protector CFLAGS option, and I'd like to change that to a Makefile `--disable-stack-protector` option.

Thanks

vincentbernat commented:
❦ 19 mai 2015 03:18 -0700, Alexandru Ardelean <notifications@github.com> :

> I do have 1 single request, if it's not too much trouble (and when you
> have time).
> It is optional.
> I've been trying to add a --disable-stack-protector option, but it
> took me too long to do it, so I gave up
> On OpenWRT, it's disabled by default (not sure why, because I do not
> maintain lldpd on OpenWRT, that's someone else). It could be that the
> toolchain (uClibc or musl) do not support stack-protection.
> Anyway, could you add it before releasing 0.7.15 ?
> I've tried lldp_ARG_ENABLE but it did not seem to work for me.
>
> Right now, on OpenWRT, there's a patch that just removes the 2
> stack-protector CFLAGS option, and I'd like to change that to a
> Makefile --disable-stack-protector option.

I think I have already asked someone in the past about this, notably if
this was possible to detect that somehow so that I automatically disable
that in configure.

But, yes, I can add an option for that. Could you open an issue so that
I don't forget?

Also, currently, it's already possible without patching. Add

ac_cv_cflags_gcc_option__fstack_protector_strong=no to configure.
-- 
Use data arrays to avoid repetitive control sequences.
            - The Elements of Programming Style (Kernighan & Plauger)

commodo commented:
Thanks :)

vincentbernat commented:
And after a Valgrind check, no memory leak. ;-)
